### PR TITLE
Express notes api

### PR DESCRIPTION
**Pull request by eniolajohnson**
image
image
image
image
image
image
image

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  if (idNum < 0 || !Number.isInteger(idNum)){
+    res.status(400).json({
+      "error": "id must be a positive integer"
+    })
+  } else if (!notes[id]){
+      res.status(404).json({
+        "error": `cannot find note with id ${id}`
+      })
+  } else if (notes[id]){
+    delete notes[id];
```
Comment (scbowler) [inline express-notes-api/index.js:59]: After you delete the note you need to write the `notes` data to the `data.json` file

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      "error": "id must be a positive integer"
+    })
+  } else if(!contentEdit){
+    res.status(400).json({
+      "error": "content is a required field"
+    })
+  } else if (!notes[id] && contentEdit){
+    res.status(404).json({
+      "error": `cannot find note with id ${id}`
+    })
```
Comment (scbowler) [inline express-notes-api/index.js:83]: Somewhere in here you need to check that `contentEdit.content` exists

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  } else if (notes[id] && contentEdit){
+    notes[id] = contentEdit;
+    contentEdit.id = id;
```
Comment (scbowler) [inline express-notes-api/index.js:86]: Once the note has been updated you need to write the `notes` data to the `data.json` file

**Review by scbowler — CHANGES_REQUESTED** [review]
Looking really good so far, but a few things need a little cleaning up. Also after any change to the `notes` data you need to write the updated `notes` object to the `data.json` file

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  } else {
+    res.json(notes[idProp]);
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  const newNote = req.body;
+  let uid = nextId;
+
+  if (JSON.stringify(newNote) === '{}'){
```
[the remark this answers is not in the document]
Comment (eniolajohnson) [reply express-notes-api/index.js]: image

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    res.status(400).json({ "error": "content is a required field" })
+  } else if (req.body) {
+    notes[uid] = newNote;
+    newNote.id = nextId++;
```
[the remark this answers is not in the document]
Comment (eniolajohnson) [reply express-notes-api/index.js:37]: image

File: express-notes-api/index.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+  if (idNum < 0 || !Number.isInteger(idNum)){
+    res.status(400).json({
+      "error": "id must be a positive integer"
+    })
+  } else if (!notes[id]){
+      res.status(404).json({
+        "error": `cannot find note with id ${id}`
+      })
+  } else if (notes[id]){
```
[the remark this answers is not in the document]
Comment (eniolajohnson) [reply express-notes-api/index.js]: image

File: express-notes-api/index.js (hunk 2 above)
Comment (eniolajohnson) [reply express-notes-api/index.js:94]: image

File: express-notes-api/index.js (hunk 3 above)
Comment (eniolajohnson) [reply express-notes-api/index.js:97]: image

File: express-notes-api/index.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    newNote.id = nextId++;
+    res.status(201).json(newNote);
+
+    let data = JSON.stringify(content, null, 2);
+
+    fs.writeFile('data.json', data, (err) => {
+      if (err) {
+        throw err
+      };
+    });
```
Comment (scbowler) [inline express-notes-api/index.js:46]: You don't want to send a response until everything has completed, including writing the data to the file. Should look something like:

```JAVASCRIPT
let data = JSON.stringify(content, null, 2);

fs.writeFile('data.json', data, (err) => {
  if (err) {
        return res.status(500).json({error: 'Unexpected server error'});
  };
  res.status(201).json(newNote);
});
```

File: express-notes-api/index.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  } else {
+    delete notes[id];
+    res.sendStatus(204);
+    let data = JSON.stringify(content, null, 2);
+
+    fs.writeFile('data.json', data, (err) => {
+      if (err) {
+        throw err
+      };
+    });
```
Comment (scbowler) [inline express-notes-api/index.js:75]: Same as above, don't send a response until everything is complete. See above comment for example.

File: express-notes-api/index.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+  if (parseInt(id) < 0 || !Number.isInteger(parseInt(id))) {
+    res.status(400).json({
+      "error": "id must be a positive integer"
+    })
+  } else if(!contentEdit.content){
+    res.status(400).json({
+      "error": "content is a required field"
+    })
+  } else if (!notes[id] && contentEdit.content){
```
Comment (scbowler) [inline express-notes-api/index.js:91]: You don't need to check the `content` property again, it was already checked above.

```JAVASCRIPT
} else if (!notes[id]) {
```

File: express-notes-api/index.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    res.status(404).json({
+      "error": `cannot find note with id ${id}`
+    })
+  } else if (notes[id] && contentEdit.content){
```
Comment (scbowler) [inline express-notes-api/index.js:95]: Don't need an `else if` here, just an `else` you have already checked both of these things they don't need checked again

File: express-notes-api/index.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    notes[id] = contentEdit;
+    contentEdit.id = id;
+    res.status(200).json(notes[id]);
+    let data = JSON.stringify(content, null, 2);
+
+    fs.writeFile('data.json', data, (err) => {
+      if (err) {
+        throw err
+      };
+    });
```
Comment (scbowler) [inline express-notes-api/index.js:105]: Same as above, don't send a response until everything is done. See above example

**Review by scbowler — CHANGES_REQUESTED** [review]
A few more changes needed, getting closer!!

File: express-notes-api/index.js (hunk 10 above)
Comment (eniolajohnson) [reply express-notes-api/index.js:95]: I put that in so I could send the `500` status. Isn't that necessary?

```
else {
    res.status(500).json({
      "error": "An unexpected error occurred."
    })
}
```

File: express-notes-api/index.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  let uid = nextId;
+
+  if (newNote.content){
+    notes[uid] = newNote;
+    newNote.id = nextId++;
+    let data = JSON.stringify(content, null, 2);
+
+    fs.writeFile('data.json', data, (err) => {
+      if (err) {
+        throw err;
```
Comment (scbowler) [inline express-notes-api/index.js:42]: Since you are on a server you should send back an error here, not throw an error

**Review by scbowler — APPROVED** [review]
Great job, looks good. Before you merge change your error handling in writeFile to send an error to the client instead of throwing an error. You can just send a 500 error:

```JAVASCRIPT
if(err) {
  return res.status(500).json({error: 'Unexpected Server Error'});
}
```

File: express-notes-api/index.js (hunk 12 above)
Comment (eniolajohnson) [reply express-notes-api/index.js]: Thank you Scott.

**Pull request merged**